cpp/lib/src/MarcRecord.cc

```diff
@@ -0,0 +1,296 @@
+#include "MarcRecord.h"
+#include "util.h"
+#include "MarcReader.h"
+
+std::string MarcRecord::getFieldData(const std::string &tag) const {
+    return getFieldData(getFieldIndex(tag));
+}
+
+
+std::string MarcRecord::getFieldData(const size_t index) const {
+    if (index == MarcRecord::FIELD_NOT_FOUND || directory_entries_.cbegin() + index >= directory_entries_.cend())
+        return "";
+    const DirectoryEntry &entry(directory_entries_[index]);
+    if (not entry.hasCache()) {
+        entry.setCache(std::string(raw_data_, entry.getFieldOffset(), entry.getFieldLength() - 1));
+    }
+    return entry.getCache();
+}
+
+
+Subfields MarcRecord::getSubfields(const std::string &tag) const {
+    return getSubfields(getFieldIndex(tag));
+}
+
+
+Subfields MarcRecord::getSubfields(const size_t index) const {
+    if (index == MarcRecord::FIELD_NOT_FOUND || directory_entries_.cbegin() + index >= directory_entries_.cend())
+        return Subfields();
+    return Subfields(getFieldData(index));
+}
+
+
+std::string MarcRecord::getTag(const size_t index) const {
+    if (index == MarcRecord::FIELD_NOT_FOUND || directory_entries_.cbegin() + index >= directory_entries_.cend())
+        return "";
+    return directory_entries_[index].getTag();
+}
+
+
+size_t MarcRecord::getFieldIndex(const std::string &field_tag) const {
+    auto const iter(DirectoryEntry::FindField(field_tag, directory_entries_));
+    return (iter == directory_entries_.end()) ? MarcRecord::FIELD_NOT_FOUND : std::distance(directory_entries_.begin(), iter);
+}
+
+
+size_t MarcRecord::getFieldIndices(const std::string &field_tag, std::vector<size_t> * const field_indices) const {
+    field_indices->clear();
+
+    size_t field_index(getFieldIndex(field_tag));
+    while (static_cast<size_t>(field_index) < directory_entries_.size() and directory_entries_[field_index].getTag() == field_tag) {
+        field_indices->emplace_back(field_index);
+        ++field_index;
+    }
+
+    return field_indices->size();
+}
+
+
+bool MarcRecord::updateField(const size_t field_index, const std::string &new_field_value) {
+    DirectoryEntry &entry(directory_entries_[field_index]);
+    size_t offset = raw_data_.size();
+    size_t length = new_field_value.length() + 1 /* For new field separator. */;
+
+    entry.setFieldLength(length);
+    entry.setFieldOffset(offset);
+    entry.setCache(new_field_value);
+    raw_data_ += new_field_value + '\x1E';
+
+    return true;
+}
+
+
+size_t MarcRecord::insertField(const std::string &new_field_tag, const std::string &new_field_value) {
+    if (new_field_tag.length() != 3)
```

Comment:
Let's use DirectoryEntry::TAG_LENGTH instead of "3" because it is sooo much nicer. ;-)  Also how about adding unlikely()?